File: core/src/main/java/xtdb/RefCounter.java
package xtdb;

import java.time.Duration;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.StampedLock;

public class RefCounter {
    private final StampedLock lock = new StampedLock();
    private AtomicInteger refs = new AtomicInteger(0);
    private volatile Semaphore closingSemaphore;

    public void acquire() {
        var stamp = lock.readLock();
        try {
            if (closingSemaphore == null) {
                refs.incrementAndGet();
            } else {
                throw new IllegalStateException("node closing");
            }
        } finally {
            lock.unlock(stamp);
        }
    }

    public void release() {
        var stamp = lock.readLock();
        try {
            refs.decrementAndGet();
            if (closingSemaphore != null) {
                closingSemaphore.release();
            }
        } finally {
            lock.unlock(stamp);
        }
    }

    public boolean isClosing() {
        return closingSemaphore != null;
    }

    public boolean tryClose(Duration timeout) throws Exception {
        int openQueries;

        var stamp = lock.writeLock();
        try {
            if (closingSemaphore != null) {
                throw new IllegalStateException("node closing");
            } else {
                openQueries = refs.get();
                closingSemaphore = new Semaphore(0);
            }
        } finally {
            lock.unlock(stamp);
        }

        if (timeout != null) {
            return closingSemaphore.tryAcquire(openQueries, timeout.toSeconds(), TimeUnit.SECONDS);
        } else {
            return closingSemaphore.tryAcquire(openQueries);
        }
    }
}

File: core/src/main/kotlin/xtdb/util/RefCounter.kt
package xtdb.util

import java.time.Duration
import java.util.concurrent.Semaphore
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.StampedLock
import kotlin.concurrent.Volatile

internal fun <V> StampedLock.readLock(f: () -> V): V {
    val stamp = readLock()
    return try {
        f()
    } finally {
        unlock(stamp)
    }
}

class RefCounter {
    private val lock = StampedLock()
    private val refs = AtomicInteger(0)

    @Volatile
    private var closingSemaphore: Semaphore? = null

    fun acquire() {
        lock.readLock {
            check(closingSemaphore == null) { "node closing" }
            refs.incrementAndGet()
        }
    }

    fun release() {
        lock.readLock {
            refs.decrementAndGet()
            closingSemaphore?.release()
        }
    }

    val isClosing: Boolean
        get() = closingSemaphore != null

    @Throws(Exception::class)
    fun tryClose(timeout: Duration?): Boolean {
        val openQueries: Int

        val stamp = lock.writeLock()

        val closingSemaphore = try {
            check(closingSemaphore == null) { "node closing" }
            openQueries = refs.get()
            Semaphore(0).also { this.closingSemaphore = it }
        } finally {
            lock.unlock(stamp)
        }

        return if (timeout != null)
            closingSemaphore.tryAcquire(openQueries, timeout.toSeconds(), TimeUnit.SECONDS)
        else
            closingSemaphore.tryAcquire(openQueries)
    }
}
